lib/src/ssh_signing.rs

```diff
@@ -0,0 +1,308 @@
+// Copyright 2023 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(missing_docs)]
+
+use std::ffi::OsString;
+use std::fmt::Debug;
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::process::{Command, ExitStatus, Stdio};
+
+use either::Either;
+use thiserror::Error;
+
+use crate::signing::{SigStatus, SignError, SigningBackend, Verification};
+
+#[derive(Debug)]
+pub struct SshBackend {
+    program: OsString,
+    allowed_signers: Option<OsString>,
+}
+
+#[derive(Debug, Error)]
+pub enum SshError {
+    #[error("SSH sign failed with exit status {exit_status}:\n{stderr}")]
+    Command {
+        exit_status: ExitStatus,
+        stderr: String,
+    },
+    #[error("Failed to parse ssh program response")]
+    BadResult,
+    #[error("Failed to run ssh-keygen")]
+    Io(#[from] std::io::Error),
+    #[error("Signing key required")]
+    MissingKey,
+}
+
+impl From<SshError> for SignError {
+    fn from(e: SshError) -> Self {
+        SignError::Backend(Box::new(e))
+    }
+}
+
+type SshResult<T> = Result<T, SshError>;
+
+fn parse_utf8_string(data: Vec<u8>) -> SshResult<String> {
+    String::from_utf8(data).map_err(|_| SshError::BadResult)
+}
+
+fn run_command(command: &mut Command, stdin: &[u8]) -> SshResult<Vec<u8>> {
+    let process = command.spawn()?;
+    process.stdin.as_ref().unwrap().write_all(stdin)?;
+
+    let output = process.wait_with_output()?;
+
+    if !output.status.success() {
+        return Err(SshError::Command {
+            exit_status: output.status,
+            stderr: String::from_utf8_lossy(&output.stderr).trim_end().into(),
+        });
+    }
+
+    Ok(output.stdout)
+}
+
+// This attempts to convert given key data into a file and return the filepath.
+// If the given data is actually already a filepath to a key on disk then the
+// key input is returned directly.
+fn ensure_key_as_file(key: &str) -> SshResult<Either<PathBuf, tempfile::TempPath>> {
+    let is_inlined_ssh_key = key.starts_with("key::") || key.starts_with("ssh-");
```

Comment:
Git appears to strip the `key::` prefix, so it's probably Git-specific thing. I don't know where this prefix comes, but maybe we can omit it for now?

Reply:
Done.